# Preparation

In [1]:
#installing pipenv
!pip install pipenv

In [7]:
#pipenv version
!pipenv --version

pipenv, version 2021.5.29



Q1 --> pipenv version is :version 2021.5.29 

In [3]:
#install scikit-learn v1.0
!pipenv install scikit-learn==1.0 --python="D:\Users\Kallel\anaconda3\python.exe" #add numpy & pandas in the next solution

#bug 1: duplicate python installation
#bug 2: 'NoneType' object has no attribute 'version_sort': googled solution: specify python full path works 
#should have been numpy + pandas

Installing scikit-learn==1.0...
Installing dependencies from Pipfile.lock (c55cee)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.


Creating a virtualenv for this project...

Pipfile: G:\repo\ML-Zoomcamp\Homework 5\Pipfile

Using d:/users/kallel/anaconda3/python.exe (3.8.8) to create virtualenv...


[    ] Creating virtual environment..
[=   ] Creating virtual environment..
[==  ] Creating virtual environment..
[=== ] Creating virtual environment..
[ ===] Creating virtual environment..
[  ==] Creating virtual environment..
[   =] Creating virtual environment..
[    ] Creating virtual environment..
[   =] Creating virtual environment..
[  ==] Creating virtual environment..
[ ===] Creating virtual environment..
[====] Creating virtual environment..
[=== ] Creating virtual environment..
[==  ] Creating virtual environment..
[=   ] Creating virtual environment..
[    ] Creating virtual environment..
[=   ] Creating virtual environment..
[==  ] Creating virtual environment..
[=== ] Creating virtual environment..
[ ===] Creating virtual environment..
[  ==] Creating virtual environment..
[   =] Creating virtual environme

* Q2 --> [sha256]  121f78d6564000dc5e968394f45aac87981fcaaf2be40cfcd8f07b2baa1e1829

Note: initially I have used the same lock file as in jupyter path, but after second iteration the lock file is generated outside, the scikit-learn checksum are matched

In [9]:
# download the files
!"C:\Program Files (x86)\TeXmacs\bin\wget.exe" https://github.com/alexeygrigorev/mlbookcamp-code/raw/master/course-zoomcamp/05-deployment/homework/dv.bin
!"C:\Program Files (x86)\TeXmacs\bin\wget.exe" https://github.com/alexeygrigorev/mlbookcamp-code/raw/master/course-zoomcamp/05-deployment/homework/model1.bin

--2021-10-06 17:26:21--  https://github.com/alexeygrigorev/mlbookcamp-code/raw/master/course-zoomcamp/05-deployment/homework/dv.bin
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/05-deployment/homework/dv.bin [following]
--2021-10-06 17:26:21--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/05-deployment/homework/dv.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 323 [application/octet-stream]
Saving to: 'dv.bin'

     0K                                                       100%  548K=0.001s

20

# Actual Work

In [4]:
#!!!! The VS Code/Jupyter kernel must be the same as virtual environment!!
# pipenv install pandas  --python="D:\Users\Kallel\anaconda3\python.exe"

import libraries, load models

In [6]:
import numpy as np
import pandas as pd
import pickle 

with open("dv.bin", 'rb') as f: dv = pickle.load(f)
with open("model1.bin", 'rb') as f: model = pickle.load(f)

## Q3 Probability a custom is churning

In [7]:
c = {"contract": "two_year","tenure": 12, "monthlycharges": 19.7} #customer
X = dv.transform([c])
prob = model.predict_proba(X)[:,1]
print("Probability that customer A is churning is %.3f" % prob)



Probability that customer A is churning is 0.115


The same results were confirmed on a virtual PC machine so it seems like 0.115 is the correct answer

## Q4: Deployment on a server

In [19]:
!pip install waitress Flask

In [22]:
# Forked to predict.py

In [4]:
# client.py: get data of the new user
import requests
url = "http://127.0.0.1:8001/predict"
customer = {"contract": "two_year", "tenure": 1, "monthlycharges": 10}
results = requests.post(url, json=customer).json()
print("Probability that a user is churning: %2.3f" % results["churn_probability"])

Probability that a user is churning: 0.999


Q4 --> 0.999

## Q5: Dockerfile's build log's hash
Q5 --> output_docker.log:: f0f43f7bc6e0

## Q6: Docker
Wasn't able to run Docker on Windows, so run it on VirtualBox, forwarded port 9696

In [9]:
url = "http://172.19.48.1:9696/predict"
customer = {"contract": "two_year", "tenure": 12, "monthlycharges": 10}
results = requests.post(url, json=customer).json()
print("Probability that a user is churning: %2.3f" % results["churn_probability"])

Probability that a user is churning: 0.329


Q6 --> 0.329